keyboards/unicomp/mini_m/keymaps/default/keymap.c

```diff
@@ -0,0 +1,80 @@
+ */
+#include QMK_KEYBOARD_H
+
+// Layer shortcuts
+enum custom_layers {
+    _QWERTY,
+    _NUMPAD,
+};
+
+enum custom_keycodes {
+    NUMSLCK = QK_USER, // Num Lock on shift, Scroll Lock regularly
+};
+
+// Mod key detection
+#define MODS_SHIFT_MASK (MOD_BIT(KC_LSFT)|MOD_BIT(KC_RSFT))
+#define MODS_ALT_MASK (MOD_BIT(KC_LALT)|MOD_BIT(KC_RALT))
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  [_QWERTY] = LAYOUT(
+    KC_ESC,           KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,             KC_PSCR, NUMSLCK, KC_PAUS,
+    KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_BSPC, KC_BSPC,   KC_INS,  KC_HOME, KC_PGUP,
+    KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC, KC_BSLS,            KC_DEL,  KC_END,  KC_PGDN,
+    KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT, KC_NUHS, KC_ENT,
+    KC_LSFT, KC_NUBS, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_RSFT, KC_RSFT,                     KC_UP,
+    KC_LCTL, KC_LGUI, KC_LALT,                            KC_SPC,                             KC_RALT, KC_RGUI, KC_APP,  KC_RCTL,            KC_LEFT, KC_DOWN, KC_RGHT
+  ),
+  [_NUMPAD] = LAYOUT(
+    KC_TRNS,          KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,            KC_TRNS, NUMSLCK, KC_TRNS,
+    KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_P7,   KC_P8,   KC_P9,   KC_TRNS, KC_PMNS, KC_PPLS, KC_TRNS, KC_TRNS,   KC_TRNS, KC_TRNS, KC_TRNS,
+    KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_P4,   KC_P5,   KC_P6,   KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,            KC_TRNS, KC_TRNS, KC_TRNS,
+    KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_P1,   KC_P2,   KC_P3,   KC_PAST, KC_TRNS, KC_TRNS, KC_PENT,
+    KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS, KC_P0,   KC_TRNS, KC_PDOT, KC_PSLS, KC_TRNS, KC_TRNS,                     KC_TRNS,
+    KC_TRNS, KC_TRNS, KC_TRNS,                            KC_TRNS,                            KC_TRNS, KC_TRNS, KC_TRNS, KC_TRNS,            KC_TRNS, KC_TRNS, KC_TRNS
+  ),
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    switch (keycode) {
+        case NUMSLCK: {
+            if (record->event.pressed) {
+                if (keyboard_report->mods & MODS_SHIFT_MASK) {
+                    tap_code(KC_NUM);
+                } else {
+                    register_code(KC_SCRL);
+                }
+            } else {
+                unregister_code(KC_SCRL);
+            }
+            break;
+        }
+    }
+    return true;
+}
+
+bool led_update_user(led_t led_state) {
```

Comment:
This should be moved to the keyboard's c file, or moved to a non-default keymap.  The default keymap should be kept plain and simple.  (same with process_record_user)

Reply:
Of course it won't work with the configurator this way :(

The Mini M, like the IBM SSK has two very inter-related features here:
1) The Numlock locklight being active activates the embedded numpad in the alpha block (so JKL=123), and those keys have secondary legends.
2) The Scroll Lock key also has a NumLock secondary legend, and pressing Shift-ScrollLock produces NumLock.

I'd like to implement this default behaviour as the default keymap, but from guidelines it sounds like this would be frowned upon?

It would involve moving `led_update_user()` to the keyboard (making layer 1 special for all keymaps, i.e. activate by numlock being on).
`process_record_user()` would also have to be moved, so the custom keycode would be defined keyboard-side. And in configurator the custom keycode would show up with the Any key, so Any(QK_USER_0), cause I suppose configurator wouldn't know my custom keycode names.

Is this something I can do, (which I'd prefer to do if it would be accepted), or should I just move this functionality in a keymap called embedded_numpad, and strip it out of the default keymap?

Reply:
Okay I have moved it to the keyboard.c file, please review.

Reply:
(Also I moved the layout macro into info.json)